Align rva fields in reflection emit

davidwrighton opened the pull request:
In support of CreateSpan (#60948), improve alignment for RVA static pre-initialized fields to align memory blocks which may contain long, ulong, or double primitive arrays on 8 byte boundaries.

Mono fix is more involved
- Fix Ref-Emit generated RVA statics
 - Set the HasFieldRVA bit. NOTE: earlier code that attempts to set the bit doesn't actually do that as the FieldRVA bit is in  FieldAttributes.ReservedMask which is masked away in the FieldBuilder constructor
 - Fix the Swizzle lookup. We should not be using the 8 byte swizzle in the final else clause.
- Enhance ref-emitted field to allow for use with CreateSpan
  - Ref-emitted fields should specify a pack size appropriate for minimum alignment of the CreateSpan targetted data
  - Respect the packing_size specified so that RVA static fields generated for CreateSpan can be properly aligned

Fixes #62314

davidwrighton committed:
Align CoreCLR System.Reflection.Emit pre-initialized data to 8 byte boundaries

davidwrighton committed:
Add test, and improve alignment check to be a bit more memory friendly

AaronRobinsonMSFT commented on src/libraries/System.Reflection.Emit/tests/ModuleBuilder/ModuleBuilderDefineInitializedData.cs:
```diff
@@ -59,5 +59,50 @@ public void DefineInitializedData_CreateGlobalFunctionsCalled_ThrowsInvalidOpera
+            // Create static field data in a variety of orders that requires the runtime to actively apply alignment
+            // RuntimeHelpers.CreateSpan requires data to be naturally aligned within the "PE" file. At this time CreateSpan only
+            // requires alignments up to 8 bytes.
+            FieldBuilder field1Byte = module.DefineInitializedData("Field1Byte", new byte[] { 1 }, FieldAttributes.Public);
```
Let's also have a test for 12 bytes, which, I believe, should result in 4-byte alignment.

davidwrighton replied:
It will result in 4 byte alignment on CoreCLR, but possibly only 2 byte on Mono. I think though, that the current set of tests is adequate.

davidwrighton committed:
Fix RVA statics, and update for CreateSpan scenario
- Enhance test case to verify that the RVA static actually has the pre-initialized data
- Fix Ref-Emit generated RVA statics
 - Set the HasFieldRVA bit. NOTE: earlier code that attempts to set the bit doesn't actually do that as the FieldRVA bit is in  FieldAttributes.ReservedMask which is masked away in the FieldBuilder constructor
 - Fix the Swizzle lookup. We should not be using the 8 byte swizzle in the final else clause.
- Enhance ref-emitted field to allow for use with CreateSpan
  - Ref-emitted fields should specify a pack size appropriate for minimum alignment of the CreateSpan targetted data
  - Respect the packing_size specified so that RVA static fields generated for CreateSpan can be properly aligned

davidwrighton commented:
@lambdageek Could you take a look. This fix has become substantially more complex in Mono.

lambdageek commented:
> @lambdageek Could you take a look. This fix has become substantially more complex in Mono.

I'm off on sick time for a bit longer - I'll try to get to it. @vargaz could you take a look too

vargaz commented:
The mono changes look ok to me.

davidwrighton committed:
Fix assertions in test

marek-safar commented:
This fails on AOT platforms. @vargaz could you have a look at the failure?
